plugins/Lv2Instrument/Lv2Instrument.cpp

```diff
@@ -0,0 +1,308 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Lv2Instrument.h"
+
+#include <QDebug>
+#include <QDragEnterEvent>
+
+#include "Engine.h"
+#include "InstrumentPlayHandle.h"
+#include "InstrumentTrack.h"
+#include "Lv2SubPluginFeatures.h"
+#include "Mixer.h"
+#include "StringPairDrag.h"
+
+#include "embed.h"
+#include "plugin_export.h"
+
+
+
+
+Plugin::Descriptor PLUGIN_EXPORT lv2instrument_plugin_descriptor =
+{
+	STRINGIFY(PLUGIN_NAME),
+	"LV2",
+	QT_TRANSLATE_NOOP("pluginBrowser",
+		"plugin for using arbitrary LV2 instruments inside LMMS."),
+	"Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>",
+	0x0100,
+	Plugin::Instrument,
+	new PluginPixmapLoader("logo"),
+	nullptr,
+	new Lv2SubPluginFeatures(Plugin::Instrument)
+};
+
+
+
+
+/*
+	Lv2Instrument
+*/
+
+
+Lv2Instrument::Lv2Instrument(InstrumentTrack *instrumentTrackArg,
+	Descriptor::SubPluginFeatures::Key *key) :
+	Instrument(instrumentTrackArg, &lv2instrument_plugin_descriptor, key),
+	Lv2ControlBase(this, key->attributes["uri"])
+{
+	if (Lv2ControlBase::isValid())
+	{
+#ifdef LV2_INSTRUMENT_USE_MIDI
+		for (int i = 0; i < NumKeys; ++i) { m_runningNotes[i] = 0; }
+#endif
+		connect(instrumentTrack()->pitchRangeModel(), SIGNAL(dataChanged()),
+			this, SLOT(updatePitchRange()), Qt::DirectConnection);
+		connect(Engine::mixer(), &Mixer::sampleRateChanged,
+			this, [this](){Lv2ControlBase::reloadPlugin();});
+
+		// now we need a play-handle which cares for calling play()
+		InstrumentPlayHandle *iph =
+			new InstrumentPlayHandle(this, instrumentTrackArg);
+		Engine::mixer()->addPlayHandle(iph);
+	}
+}
+
+
+
+
+Lv2Instrument::~Lv2Instrument()
+{
+	Engine::mixer()->removePlayHandlesOfTypes(instrumentTrack(),
+		PlayHandle::TypeNotePlayHandle |
+						  PlayHandle::TypeInstrumentPlayHandle);
+}
+
+
+
+
+bool Lv2Instrument::isValid() const { return Lv2ControlBase::isValid(); }
+
+
+
+
+void Lv2Instrument::saveSettings(QDomDocument &doc, QDomElement &that)
+{
+	Lv2ControlBase::saveSettings(doc, that);
+}
+
+
+
+
+void Lv2Instrument::loadSettings(const QDomElement &that)
+{
+	Lv2ControlBase::loadSettings(that);
+}
+
+
+
+
+void Lv2Instrument::loadFile(const QString &file)
+{
+	Lv2ControlBase::loadFile(file);
+}
+
+
+
+
+#ifdef LV2_INSTRUMENT_USE_MIDI
+bool Lv2Instrument::handleMidiEvent(
+	const MidiEvent &event, const MidiTime &time, f_cnt_t offset)
+{
+	// this function can be called from GUI threads while the plugin is running,
+	// so this requires caching, e.g. in ringbuffers
```

Comment:
I never thought of fixing this in the core, I thought every plugin would have its own ringbuffer. Having it in the core sounds generic, which is good. I guess it would not be a virtual `handleMidiEvent` then anymore, but rather a `getMidiEventFromRingbuffer` function that the plugins would call in the realtime thread?

Reply:
Ah, I must have misunderstood what you meant there. I was thinking that any note input not on the realtime thread (e.g. mouse clicks on a piano view or computer keyboard input which arrive on the UI thread, or MIDI device input which arrives on a MIDI callback thread) immediately goes into a ringbuffer. This is read by the realtime thread every processing cycle, and the appropriate MIDI messages are dispatched from there, along with all those from pattern playback, via the current system.

This should address a few threading issues; not just instrument ones such as #5015, but also core ones such as #5373. It will also make MIDI processing more symmetrical with play handle processing, which has always been on the realtime thread only.

It's also worth noting, however, that there are multiple realtime threads in LMMS - one per core - so instruments may still receive events from multiple threads simultaneously even if they're not on the UI thread. I think this should be changed too so that each instrument is only handled by one of the realtime threads - it's the cause of bugs such as #2606 (ignore the title there - it's currently incorrect and misleading), complicates the implementation of instruments like this, and will make it difficult to implement support for MIDI effect plugins.

This all deserves its own issue though, so I think you can keep the comment for now and we'll see what gets implemented first - threading fixes or LV2 instruments.

Reply:
I'm not sure what you meant by "immediately" in the first paragraph, but I think I can imagine. In the end, ringbuffers must be at least at the intersection between realtime and non realtime threads, so e.g. if you press a "live" note on the piano roll, that's non-realtime and the MIDI event must "somewhere" be stored in a buffer, wherever that is. I'm not sure e.g. how to handle notes from the piano roll. They are actually fixed, but I assume you can move a note while the song is playing.

We currently need to provide this callback, though it's currently not RT safe. If we will fill it, we must consider a ringbuffer to not destroy the Lv2 realtime conditions. Maybe the ringbuffers will later be in the core and writing LMMS plugins gets more comfortable. But for now, having a ringbuffer reminding comment there is a defensive measure.

Reply:
> I'm not sure what you meant by "immediately" in the first paragraph

I meant "without any further processing" - i.e. it's all handled on a realtime thread.

> Maybe the ringbuffers will later be in the core and writing LMMS plugins gets more comfortable. But for now, having a ringbuffer reminding comment there is a defensive measure.

👍